### DISCUSSION TITLE
MUIRichTextEditor integration

### DISCUSSION THREAD

**Discussion by keemor** [react-hook-form/react-hook-form#1699] [General] [answered]
Hi,

I'm trying to integrate MUIRichTextEditor [mui-rte](https://github.com/niuware/mui-rte) with react-hook-form

[codesandbox.io/s/react-hook-form-controller-c0mr0](https://codesandbox.io/s/react-hook-form-controller-c0mr0?file=/src/index.js)

1. Attempt, value is not passed to handleSubmit and if turn on defaultValue="RTE defaultValue" 
there's an error _A cross-origin error was thrown._
```
<MUIRichTextEditor
  //Line below throwns error => A cross-origin error was thrown.
  //  defaultValue="RTE defaultValue"
  label="Test"
  name="RTE2"
  ref={register}
/>
```
2. Attempt throws the same error _A cross-origin error was thrown_ when I start typing inside textarea
```
<Controller
  as={<MUIRichTextEditor label="Test" />}
  name="RTE1"
  control={control}
/>
```

I'm aware that it might be MUIRichTextEditor problem.

Maybe there's some other RTE which can be easily integrated with hook-form?

Cheers,

**Comment by bluebill1049** [comment, marked as answer]
can you leverage `onChange` prop at the `Controller`.

https://react-hook-form.com/api#Controller

**Reply by keemor** [reply]
Thanks @bluebill1049 for the tip, this is how it works:
```
<Controller
  as={MUIRichTextEditor}
  name="RTE1"
  label="Start typing"
  control={control}
  onChange={([value]) => {
	const content = JSON.stringify(
	  convertToRaw(value.getCurrentContent())
	);
	return content;
  }}
/>
```

[codesandbox.io/s/react-hook-form-muirichtexteditor-integration-c0mr0](https://codesandbox.io/s/react-hook-form-muirichtexteditor-integration-c0mr0)

**Reply by zszep** [reply]
If you type text into your example on codesandbox, the text is being entered in reverse. Have the same problem in my code. Any solution for this?

**Reply by bluebill1049** [reply]
here is the alternative version with custom register:

https://codesandbox.io/s/react-hook-form-muirichtexteditor-integration-4zgse?file=/src/index.js

**Reply by zszep** [reply]
Great, this works. There is one problem though. I cannot set an initial value to the textbox. I guess it is because MuiRichTextbox does not support a "name" attribute.  Any solution to this?

**Reply by bluebill1049** [reply]
you will have to use their defaultValue prop

**Reply by zszep** [reply]
Great support. Works like a charme now.

**Reply by bluebill1049** [reply]
np

**Reply by nonoumasy** [reply]
is it possible to get an update for this with RHF version 7? Cheers

**Reply by francodgstn** [reply]
Hi, thanks for the solution. I think there's still a little issue, updating the value doesn't make the form dirty.
👉 https://codesandbox.io/s/react-hook-form-muirichtexteditor-integration-forked-vbznf7 

**UPDATE**
Sorry, I didn't see that it can be forced by setting shouldDirty to true.
`setValue("RTE1", content, { shouldDirty: true })`

**Comment by wowczarczyk** [comment]
In the codesandbox provided when I type something in the RTE, the cursor is moved to the beginning (as if render occured though the render count is not incremented)

**Reply by bluebill1049** [reply]
it's working as expected to me. i just tried. what's your steps to reproduce the issue?

**Reply by wowczarczyk** [reply]
MacOS Catalina
Chrome Version 83.0.4103.97 (Official Build) (64-bit) (Guest window to avoid extensions inference) 
I open sandbox: https://codesandbox.io/s/react-hook-form-muirichtexteditor-integration-c0mr0
I start typing in the RTE field with following keystorkes: 'a','b','c','d'. The RTE field displays: dcba

![image](https://user-images.githubusercontent.com/1281649/84747637-ce1b6800-afb7-11ea-8c4a-1a1784385a60.png)

**Reply by wowczarczyk** [reply]
Same happens on Safari and Firefox

**Reply by bluebill1049** [reply]
i will take a look this issue.

**Reply by bluebill1049** [reply]
not quite familiar with this component and did some investigation. I think it would probably easier to use custom `register` at `useEffect` and update value with `setValue`.

**Reply by bluebill1049** [reply]
i think they are going away with controlled, i can see `value` is been deprecated as well. (in the doc)

**Reply by bluebill1049** [reply]
https://codesandbox.io/s/react-hook-form-muirichtexteditor-integration-4zgse?file=/src/index.js

**Reply by wowczarczyk** [reply]
Hey, yes, this works fine, thank you!

**Comment by satinder1624** [comment]
All the links taking me to the same codeSandbox. https://codesandbox.io/p/sandbox/react-hook-form-muirichtexteditor-integration-c0mr0?file=%2Fsrc%2Findex.js

I am looking to use Controller with RTF

**Comment by satinder1624** [comment]
So, let me trying to explain my problem. I am using rich text field with react hook from. I am not able to update the value correctly. I fI uncomment setValue and type abcd it will appear dcba

<Controller
                        render={({field: {onChange, value}}) => (
                            <MUIRichTextEditor
                                onChange={value => {
                                    const content = stateToHTML(value.getCurrentContent())
                                    // setValue('comment', content)
                                    return content;
                                    // console.log('Hey ->', content)
                                }}
                                inlineToolbar
                                inheritFontSize
                                defaultValue={htmlToDraftBlocks(value)}
                                controls={['bold', 'italic', 'underline']}
                                label={'Comment'}
                                id={'comment'}
                            />
                        )}
                        name="comment"
                        label="Comment"
                        control={control}
                    />
